app/src/main/java/com/abominableshrine/taptounlock/UnlockService.java

```diff
@@ -0,0 +1,218 @@
+package com.abominableshrine.taptounlock;
+
+import android.app.Service;
+import android.content.BroadcastReceiver;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.content.ServiceConnection;
+import android.os.IBinder;
+import android.os.Messenger;
+import android.util.Log;
+import android.view.KeyEvent;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+/**
+ * The Service that can lock and unlock the device. It does so if it receives a message from
+ * a client that tells it to do so.
+ */
+public class UnlockService extends Service {
+    /* *****************************
+     * SECTION 1: GLOBAL VARIABLES
+     * ***************************** */
+    private final static Boolean DEBUG = AppConstants.DEBUG;
+    private boolean mBound;
+    private static Boolean running; // Boolean class to enable check for null
+
+    protected static UnlockService instance;
+
+    // Variables used for the communication to the TapPatternDetectorService
+    private Intent tapPatternDetectorServiceIntent;
+    private IBinder mIBinder;
+    private Messenger mUnlockServiceMessenger = null;
+    private ServiceConnection mUnlockServiceConnection = new ServiceConnection() {
+        public void onServiceConnected(ComponentName className, IBinder service) {
+            mUnlockServiceMessenger = new Messenger(service);
+            mBound = true;
+            mIBinder = mUnlockServiceMessenger.getBinder();
+        }
+
+        public void onServiceDisconnected(ComponentName className) {
+            mUnlockServiceMessenger = null;
+            mBound = false;
+            mIBinder = null;
+        }
+    };
+    private TapPattern mTapPattern = new TapPattern();
+    // The client class is used to enable simple communication to the TapPatternDetectorService
+    private UnlockServiceTapPatternDetectorClient mUnlockServiceTapPatternDetectorClient;
+
+    // Variables used for the ScreenOffBroadcastReceiver
+    private static ScreenOffBroadcastReceiver mScreenOffBroadcastReceiver;
+    private IntentFilter screenOffFilter;
+
+    /* *****************************
+     * SECTION 2: METHODS
+     * ***************************** */
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        instance = this;
+        running = true;
+
+        // Initialize communication to the TapPatternDetectorService
+        tapPatternDetectorServiceIntent = new Intent(this, TapPatternDetectorService.class);
+        bindService(tapPatternDetectorServiceIntent, mUnlockServiceConnection, Context.BIND_AUTO_CREATE);
+        mUnlockServiceTapPatternDetectorClient = new UnlockServiceTapPatternDetectorClient(mIBinder);
```

Comment:
Your mBinder will be `null` until [this line](https://github.com/psycholold/TapToUnlock/pull/6/files#diff-4643a32b4c51a9a0403c4542b8ac2332R40) gets executed. Since the service is bound asynchronously you don't know when this will happen. Therefore you can not guarantee that `mBinder` is valid by the time you create the client if you don't do it in the callback.